DataFormats/Portable/interface/PortableDeviceProduct.h

```diff
@@ -0,0 +1,72 @@
+#ifndef DataFormats_Portable_interface_PortableDeviceProduct_h
+#define DataFormats_Portable_interface_PortableDeviceProduct_h
+
+#include <cassert>
+#include <optional>
+#include <type_traits>
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+
+// generic SoA-based product in device memory
+template <typename T, typename TDev, typename = std::enable_if_t<alpaka::isDevice<TDev>>>
+class PortableDeviceProduct {
```

Comment:
What is the added value of this class compared to plain Alpaka buffer?

I can think of
* can be default-constructed
* handles the data format `const`-requirements of the framework correctly

Are these correct? Is something missing?

(I'm mostly trying to understand the situation better, e.g. if we could eventually find a way without this level of indirection)

Reply:
How about `PortableDeviceObject` as the name? The name "product" is quite specific to how framework handles "data products", whereas this class seems to address a bit different concern. Also `edm::DeviceProduct<PortableDeviceProduct<Foo>>` looks repetitive with the name "Product".

Reply:
And for the "general one" would go with `PortableObject` instead of `PortableProduct`.